### ISSUE TITLE
SQL Server: consider adding first class support for SqlConnection.AccessToken

### ISSUE THREAD

**Issue by divega** [aspnet/EntityFrameworkCore#13261]
SqlConnection.AccessToken is added to SqlClient for .NET Core in 2.2 to enable Azure Active Directory Authentication. 

Although it is already possible to use the feature with the current APIs in EF Core and our SQL Server provider, it requires creating the SqlConnection object and passing it to UseSqlServer (or somehow setting the access token just before the connection object is ever opened). 

That pattern is a bit more convoluted and forces you to deviate from the common pattern in ASP.NET Core applications that uses AddDbContext with a simple connection string. 

This issue is about making this more first-class, for example, by enabling passing the AccessToken in the UseSqlServer call as a separate piece of configuration from the connection string. 

We should understand the common patterns to use the feature safely to decide what APIs we actually need:

- Is the access token obtained using async API?
- Can we assume that the authentication token will be obtained only once and then can be cached?

**Comment by Appjunkie** [comment]
@divega Do you have a code sample that I can refer? 
I don't quite get the approach described in the link.

**Comment by divega** [comment]
@Appjunkie sorry, I don't have an example at the moment. I need to work on one, but I am not sure when I will get to it. Perhaps the original poster of the workaround can help. cc @cbrianball.

**Comment by cbrianball** [comment]
@Appjunkie I posted some [sample code](https://github.com/aspnet/EntityFrameworkCore/issues/11928#issuecomment-454853868)

**Comment by ajcvickers** [comment]
See also #15247 and consider:
* When/where/how frequently the call to get an access token should be made
* Allowing this to be done asynchronously

**Comment by ChristopherHaws** [comment]
I would love to see this added as a supported feature. In the meantime, I am doing this which seems to work pretty well: https://gist.github.com/ChristopherHaws/b1c54b95838f1513bfb74fa1c8e408f3
Ideally, we wouldn't even need to do this if we could specify something in the connection string and have SqlClient handle it itself.

**Comment by svrooij** [comment]
@ChristopherHaws even with your very nice solution I'm getting an error `System.Data.SqlClient.SqlException (0x80131904): Login failed for user 'NT AUTHORITY\ANONYMOUS LOGON'.`. Could you show us the used connection string? I think that might be where I made a mistake. Mine looks like
```
Server=tcp:correct-server.database.windows.net,1433;Initial Catalog=existingdb;Persist Security Info=False;MultipleActiveResultSets=False;Encrypt=True;TrustServerCertificate=False;
```

**Comment by cbrianball** [comment]
@svrooij It's likely not a connection string issue. The error message you gave is, unfortunately, a very generic one that doesn't help (as is usually the case with security-based exceptions). It likely means the user you are trying to connect with doesn't have permissions to connect to the database.

Assuming you're using an MSI (or user-assigned service identity), and you've configured permissions correctly, then it could simply be a waiting game as permissions propagate (I've had to wait hours before it would work in the past).

If you haven't already, I would recommend using a [connection string](https://docs.microsoft.com/en-us/azure/key-vault/service-to-service-authentication#connection-string-support) (separate from the database connection string) for AzureServiceTokenProvider. The reason I recommend this is that if you don't, the API will essentially try each of it's own internal providers until it finds one that works, this added 7+ seconds to obtaining a token in the past.

**Comment by svrooij** [comment]
Thank for the tip, I was already using that. Locally it’s set to `RunAs=Developer; DeveloperTool=VisualStudio` and in Azure it’s set to `RunAs=App`. Locally it fetches a token from azure that shows my user ID, has the correct audience and includes the guids of the Azure AD groups.
What claims should be in the token to be valid for SQL?
How can I debug this? At the moment it seems to be a black box to me. Already tried setting the logging for `Microsoft.EntityFrameworkCore` to `Debug` but that doesn’t help much.

**Comment by cbrianball** [comment]
@svrooij Sounds like you are doing everything right. If you are using MSI (instead of user-assigned identity) and you want to see the actual token, you can use the following Power Shell script and execute it on the VM/App Service to output the token (you will want to update the resource query string parameter). If you are using an App Service, I use the App Service editor (there's an 'Open Console' menu item) to execute the Power Shell.

```
$uri = "$($env:MSI_ENDPOINT)?resource=https://graph.microsoft.com/&api-version=2017-09-01"
$ProgressPreference = "SilentlyContinue"	
# Get an access token for managed identities for Azure resources
$response = Invoke-WebRequest -UseBasicParsing -Uri $uri ` -Headers @{Secret="$env:MSI_SECRET"}
$responseAsString = [System.Text.Encoding]::ASCII.GetString($response.Content)
echo $responseAsString
```

I'm not sure what would need to change in the script to get it to work with a user assigned identity (I've never used that feature before).

**Comment by syndicatedshannon** [comment]
> In the meantime, I am doing this which seems to work pretty well: https://gist.github.com/ChristopherHaws/b1c54b95838f1513bfb74fa1c8e408f3
 - https://github.com/aspnet/EntityFrameworkCore/issues/13261#issuecomment-485025402

@ChristopherHaws Hey, Chris, can you comment on why you override `CreateMasterConnection`? Is that a source of problems in some scenarios? If so, then other recommended approaches (such as the following) don't account for this:
 - https://github.com/aspnet/EntityFrameworkCore/issues/11928#issuecomment-455312550

Side note, I also found @roysanchez [implementation](https://gist.github.com/roysanchez/b06183068d303a535840e557efbccadb), which adds `IsMultipleActiveResultSetsEnabled` and `SupportsAmbientTransactions`. These may be Azure-specialized values of interest to other reading this thread as well, I'm not sure.

**Comment by syndicatedshannon** [comment]
Side note, I also found [this implementation](https://gist.github.com/roysanchez/b06183068d303a535840e557efbccadb) from @roysanchez, which adds `IsMultipleActiveResultSetsEnabled` and `SupportsAmbientTransactions`. These may be Azure-specialized values of interest to other reading this thread as well, I'm not sure.

**Comment by ChristopherHaws** [comment]
@syndicatedshannon I need to override it because it needs to return a new instance of itself. The default implementation returns a new `SqlServerConnection` and the overridden one returns an `AzureSqlServerConnection`. Without overriding it, `CreateMasterConnection` won't authenticate with an access token. I agree, I don't like overriding it either. I am hoping for a built in solution soon so I can remove my custom implementation of a pubternal type. :)

**Comment by syndicatedshannon** [comment]
Related comment on the necessity of patching `CreateMasterConnection`: https://github.com/MicrosoftDocs/azure-docs/issues/38814#issuecomment-531753854

**Comment by syndicatedshannon** [comment]
An alternative:
https://github.com/Azure-Samples/app-service-msi-entityframework-dotnet

was announced here in April 2019:
https://azure.microsoft.com/en-us/blog/securing-azure-sql-databases-with-managed-identities-just-got-easier/

**Comment by cbrianball** [comment]
@syndicatedshannon Thanks for pointing this out. The second link you provided would be nice, but as you suspected, it is currently only available for .NET Framework. The base class that it relies on does not exist in .NET Standard (or .NET Core); however, Microsoft has released a new API for SQL (Microsoft.Data.SqlClient), and it DOES have the base class that the second link relies on, so even if Microsoft.Azure.Services.AppAuthentication doesn't implement it for .NET Core, you could still roll your own -- I doubt it would be complicated to do so.

For now my app is a .NET Core 2 app, and I use EF Core, which relies on System.Data.SqlClient (it may be possible to switch it out, I don't know), but when I upgrade to .NET Core 3.0, (and EF Core along with it), then I will definitely look into this. Thanks for posting!

**Comment by syndicatedshannon** [comment]
Hah! Thanks for the verification. Funny, I hadn't looked any further yesterday, but I just went back today and noticed:

`#if net472

https://github.com/Azure/azure-sdk-for-net/blob/0acf2f02c648407da123a3a3b9a352041f7ddd4d/sdk/mgmtcommon/AppAuthentication/Azure.Services.AppAuthentication/SqlAppAuthenticationProvider.cs#L4

**Comment by syndicatedshannon** [comment]
Nice FAQ on that library, which includes a link to SqlClient repo: https://github.com/dotnet/SqlClient/wiki/Frequently-Asked-Questions

**Comment by syndicatedshannon** [comment]
The NetFx library worked like a charm after setting to build for NetCore against the latest stable Microsoft.Data.SqlClient library. Here are details:

1) Build the https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/mgmtcommon/AppAuthentication/Azure.Services.AppAuthentication/ library against Microsoft.Data.SqlClient instead of System.Data.SqlClient. Note you'll have to remove or change the `#if net472` preprocessor directive, and add a netstandard2.0 target framework.

2) The connection string must be edited as described above. The 'SqlAppAuthenticationParameters' class depends on three values: Authority, Resource, UserId. Authority means you must include a server domain name that supports SqlTokenAuthentication. For me, it ends with ".database.windows.net". Also include the following connection string parameters:
`UID=AnyString;Authentication=Active Directory Interactive`

3) In place of the web.config described above, register your custom SqlAppAuthenticationProvider class as the static provider for SqlAuthenticationMethod.ActiveDirectoryInteractive. You can do that with the following, probably clearest just before your UseSqlServer call:
`SqlAuthenticationProvider.SetProvider(SqlAuthenticationMethod.ActiveDirectoryInteractive, new SqlAppAuthenticationProvider());`

**Comment by syndicatedshannon** [comment]
Also, remove all of the other hacks listed above.

**Comment by syndicatedshannon** [comment]
https://github.com/syndicatedshannon/azure-sdk-for-net/commit/57e3dd6fea23d2cefcf04508dcbc5f88b7d221a0

**Comment by OskarKlintrot** [comment]
@syndicatedshannon I try to get a `SqlAuthenticationProvider` initialized using your example line of code in `Startup.cs`:
```csharp
SqlAuthenticationProvider.SetProvider(SqlAuthenticationMethod.ActiveDirectoryInteractive, new FakeSqlAuthenticationProvider());
```

`FakeSqlAuthenticationProvider` returns `true` in `IsSupported()` and just throws a `NotImplementedException` in `AcquireTokenAsync()` just to see if it gets used at all before I try to learn what I can use it for. However neither using `new SqlConnection(connectionString)` or EF Core triggers the method `AcquireTokenAsync()`. Am I missing something?

**Comment by syndicatedshannon** [comment]
@OskarKlintrot Sorry, I don't remember the details, and I got my hands full on another thing. I'll be getting back to this issue in about a week, sooner if I'm lucky or good. Will follow up here when I do.

**Comment by syndicatedshannon** [comment]
Short story is, I think it wasn't behaving as I expected either, and I meant to get back to it. This PR is still outstanding in our pipeline. Sorry if I led you astray.

**Comment by OskarKlintrot** [comment]
No worries, I used @ChristopherHaws [gist](https://gist.github.com/ChristopherHaws/b1c54b95838f1513bfb74fa1c8e408f3#file-azuresqlserverconnection-cs) as inspiration to get it working instead.

**Comment by ajcvickers** [comment]
Consider renewal scenarios: #19808

**Comment by sebader** [comment]
@ajcvickers do you have any update on the roadmap for this? Would be really great to see first class support for this to get away from workaround like the one I'm using from @ChristopherHaws

**Comment by ajcvickers** [comment]
@sebader This issue is in the backlog with `consider-for-next-release`. This means it is currently a stretch goal for 5.0. See [release planning](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning).

**Comment by sebader** [comment]
got it! Fingers crossed then. The easier it gets for people to get away from traditional connection strings with passwords etc, the better :)

**Comment by syedhassaanahmed** [comment]
@ajcvickers Any update on this issue? We'd like to use EF Core with Managed Identity Auth to SQL Server and this one is blocking us.

**Comment by OskarKlintrot** [comment]
@syedhassaanahmed Have you seen @ChristopherHaws [gist](https://gist.github.com/ChristopherHaws/b1c54b95838f1513bfb74fa1c8e408f3)? I have been using EF Core with Managed Identity for a few months now. I used `IDbContextOptionsExtension` though, so I did not follow the gist exactly. But it should be working. This issue should not be blocking you.

**Comment by syedhassaanahmed** [comment]
Thank @OskarKlintrot I'm looking into it right now. Its a bit unclear how does it handle renewals when the token expire and existing connection lives. Did you manage to solve that part? Do you have an example?

**Comment by OskarKlintrot** [comment]
The `AzureServiceTokenProvider` renews the token 5 min before it expires and I'm using the DbContext in a WebApi so it's registered as scoped. So for me it's not a problem.

**Comment by syedhassaanahmed** [comment]
We're also using DbContext in web so its not a problem there. For DbContext in long running jobs however, its still an issue.

**Comment by OskarKlintrot** [comment]
You could use [`IDbCommandInterceptor`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.diagnostics.idbcommandinterceptor?view=efcore-3.1) to be to renew it if needed everytime you try to access the database. The token is [cached in a static field](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/mgmtcommon/AppAuthentication/Azure.Services.AppAuthentication/Cache/AppAuthResultCache.cs#L14) so there should not be much of a performance penelty if the cache hasent expired yet. If it has though retriving a new token once an hour during a long running job might not matter that much anyway.

**Comment by ajcvickers** [comment]
See also #20969 which is a similar issue for Cosmos.

**Comment by AndriySvyryd** [comment]
We might not need to do anything here after https://github.com/dotnet/SqlClient/pull/730

**Comment by sopelt** [comment]
Looks nice for production ... a solution using Azure.Identity would avoid having to use a completely different approach for local development.

**Comment by sebader** [comment]
@sopelt I think you might have a good point there. Probably worth for you to comment on that PR as well, not just here

**Comment by jabbera** [comment]
The MSI solution only runs on Azure based infrastructure. If arc supported MSI I would be a buyer, but you still can't do local development against and azure database easily which makes me sad.

**Comment by ericsampson** [comment]
yeah dotnet/SqlClient#730 is great, but will still need to figure out a solution for localdev.

**Comment by roji** [comment]
For anyone here, [this blog post](https://devblogs.microsoft.com/azure-sdk/azure-identity-with-sql-graph-ef/#ef-core-integration) shows how to write a trivial interceptor to use [Azure Identity](https://www.nuget.org/packages/Azure.Identity).

Note to implementor: any solution we'd provide out of the box in EF Core would like make use of that library as well under the hood. This would have the consequence of taking a reference on Azure Identity from the SQL Server provider, which isn't ideal for everyone not using Azure. Some sort of plugin may mitigate that, but I'm not sure we currently have the proper extension point for that.

**Comment by ErikEJ** [comment]
@roji M.D.S. already depends on https://www.nuget.org/packages/Microsoft.Identity.Client/

**Comment by roji** [comment]
Yeah, in that case no worries to add that dependency at the EF Core level, thanks!

**Comment by roji** [comment]
Though am not sure what the exact relation is between Microsoft.Identity.Client and Azure.Identity... Seems that the latter depends on the former, so we may need to add an additional dependency (which isn't necessarily a problem).

**Comment by ErikEJ** [comment]
@roji Some additional information about Azure.Identity and SqlClient https://github.com/dotnet/SqlClient/pull/730#issuecomment-700872728 (probably whole discussion in PR is of interest)

**Comment by roji** [comment]
Thanks @ErikEJ!

**Comment by roji** [comment]
**tl;dr** fetching the access token on each and every connection (i.e. calling GetTokenAsync) is a perf anti-pattern for at least some scenarios, since tokens aren't always internally cached by Azure.Identity.

There are currently various samples for setting up an interceptor to set the access token on SqlConnection before opening it ([see this great blog post](https://devblogs.microsoft.com/azure-sdk/azure-identity-with-sql-graph-ef/#ef-core-integration), as well as [this docs PR sample](https://github.com/dotnet/EntityFramework.Docs/pull/2787#discussion_r507088960)). I've investigated if the pattern shown by these resources is OK from a perf perspective, and got the following response from @schaabs:

> While many of the credential types in Azure.Identity do in fact cache tokens, not all do, and if you're directly calling GetTokenAsync the best practice is to cache the tokens returned, especially if they are in a hot code path. We've recently updated our documentation around GetTokenAsync to provide better guidence here, adding the following.
>
>> This method is called automatically by Azure SDK client libraries. You may call this method directly, but you must also handle token caching and token refreshing.
>
> In this case the example from the blog post is using the EnvironmentCredential and the ManagedIdentityCredential. The EnvironmentCredential will cache tokens it obtains, but the ManagedIdentityCredential will not.  The ManagedIdentityCredential does make a network call, but it's a loop back call to a local endpoint so the latency wouldn't be terrible. However, if OpenAsync is called with high frequency it's possible that calls will get throttled. We should update the blog post to cache the acquired tokens and only call GetTokenAsync when the token is near expiration.

So interceptors (as well as any feature developed here) should be configured with a user-provided timeout, during which they cache the access token.

[Note this very useful discussion](https://github.com/dotnet/SqlClient/issues/767) on how access tokens are handled between SqlClient and SQL Server.

**Comment by ericsampson** [comment]
@roji, I'm a little confused - I thought we'd just be able to use the new connection string formats added by SqlClient; are you planning on adding Managed Identity support in EFCore using Azure.Idendity instead (or in addition)?

**Comment by roji** [comment]
@ercisampson if you just want to do stuff via the connection string, there's nothing you need from EF Core here - you can already pass an access token that way. Unless I'm missing something, this is in order to support setting the new SqlConnection.AccessToken property, programmatically.

Importantly, a full solution here needs to support rotation for the access tokens. Once again, if you want to do everything via the connection string, that means it's your responsibility as a user to change that connection string, integrating a new connection string before the current one expires. A more turn-key solution would automatically take care of getting a new access token when necessary, and assigning that to SqlConnection.AccessToken without the user needing to do anything.

Does that make sense?

**Comment by roji** [comment]
Note https://github.com/dotnet/SqlClient/issues/771, which is about implementing the bulk of this inside SqlClient, making the Azure.Identity integration available to non-EF Core users as well. Assuming that's done, this issue can track a thin wrapping around that support in the EF SqlServer provider.

**Comment by ericsampson** [comment]
@roji , I'm not sure if this is what you're looking for, but SqlClient 3.0 now uses Azure.Identity and adds support for the equivalent of DefaultAzureCredentials:
https://github.com/dotnet/SqlClient/blob/main/release-notes/3.0/3.0.0.md#active-directory-default-authentication-support

**Comment by ErikEJ** [comment]
@ericsampson 3.0 is not LTS, so adding that to EF Core will happen with EF Core 7.

**Comment by ericsampson** [comment]
3.0 of SqlClient? I didn't realize that package had the concept of LTS or not.

Thanks Erik :)

**Comment by ErikEJ** [comment]
It does, and EF Core 6 LTS needs LTS dependencies.

**Comment by sander1095** [comment]
> 
> 
> @ericsampson 3.0 is not LTS, so adding that to EF Core will happen with EF Core 7.

@ErikEJ 

Hello Erik!

Currently EF Core Sql Server uses Microsoft.Data.SqlClient 2.0.1 but I couldn't get the `Authentication` key in my connectionstring to work (for Managed Identity auth) even though 2.0 should support this if I understand correctly. 

Not to worry, I just installed 3.0 manually and now everything works.

If I understand correctly, you mean that only EF Core 7 will update to 3.0, not EF Core 6.0? Will 6.0 perhaps update to v2.1 where the `Authentication` property might work better for my scenario?

Also, do you have a source for Microsoft.Data.SqlClient 3.0 not being LTS? I'd love to read more about this.

Thanks!

**Comment by ErikEJ** [comment]
@sander1095 I think you just need SqlClient 2.1 as documented here: https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/azure-active-directory-authentication?view=sql-server-ver15#setting-azure-active-directory-authentication 

6.0 uses 2.1, yes: https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/EFCore.SqlServer.csproj#L24 

Sure, the SqlClient support matrix is here: https://docs.microsoft.com/en-us/sql/connect/ado-net/sqlclient-driver-support-lifecycle?view=sql-server-ver15
